### PR TITLE
ECR cleanup code review & overhaul

### PR DESCRIPTION
**Pull request by rclark**
@arunasank I made a lot of changes to the existing code from #97. Wanted to open it up as a PR to provide some insight into why I made this set of changes. See below.

### CODE REVIEW

File: scripts/cleanup.js [hunk 1]
```diff
@@ -53,124 +67,116 @@ function getImages(region, repo, callback) {
...
+      if (result.type === 'generic')
+        generic.push({ imageDigest: result.digest });
+      if (result.type === 'priority')
+        priority.push({ imageDigest: result.digest });
+    });
 
-  });
+    // We want to leave the repository with MAX_IMAGES number of images in it.
+    let imagesToDelete;
+    let excessImages = images.length - MAX_IMAGES;
```
Comment (rclark) [inline scripts/cleanup.js]: These two variables are god examples of where you should use `let`. You should prefer `const` over let if you are not going to actually change the reference.

Further along here, you see that there are several statements like

```
imagesToDelete = some new object for this variable to reference
```

```
excessImages = excessImages - some value
```

Only when you're changing what the variable will reference by setting it `=` a new thing should you use `let`.

File: scripts/cleanup.js [hunk 2]
```diff
@@ -53,124 +67,116 @@ function getImages(region, repo, callback) {
...
+  const ecr = new AWS.ECR({ region: region });
 
-  let ecr = new AWS.ECR({ region: region });
-  ecr.batchDeleteImage({
-    imageIds: imageIds,
-    repositoryName: repo
-  }, callback);
+  // Must delete images in batches of 100 max
+  const remaining = JSON.parse(JSON.stringify(images));
+  const imageIds = remaining.splice(0, 100);
```
Comment (rclark) [inline scripts/cleanup.js]: Needed to account for the limit in number of images that can be deleted in one request.

File: scripts/cleanup.js [hunk 3]
```diff
@@ -53,124 +67,116 @@ function getImages(region, repo, callback) {
...
-    cmd.stdout.on('data', out => {
-      stdout = out.toString('utf-8').trim();
-    });
-    cmd.on('close', () => {
-      // An error is just the git command erroring, which means that the
-      // output of the current command is void for determining type, as
-      // opposed to a true error in the script. So, just return null,
-      // instead of erroring.
-      if (stderr && stderr.length) return callback(null, null);
-      else return callback(null, stdout);
```
Comment (rclark) [inline scripts/cleanup.js]: You've written here a `spawn` wrapper that is essentially the same as `exec`. See [node.js docs](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback).

The great thing about `exec` over `spawn` is that you get to just write a shell command, complete with pipes and whatever else. You also get `stdout` and `stderr` returned to you for free -- no need to collect it yourself.

File: scripts/cleanup.js [hunk 4]
```diff
@@ -53,124 +67,116 @@ function getImages(region, repo, callback) {
...
+    child_process.exec(command, (err, stdout) => {
+      if (err) return callback();
+      callback(null, stdout.trim());
     });
-  } 
+  }
+
+  const merge = `git --git-dir=${tmpdir}/.git cat-file -p ${sha} | grep -Ec '^parent [a-z0-9]{40}'`;
+  const tag = `git --git-dir=${tmpdir}/.git tag | grep ${sha}`;
+  const commit = `git --git-dir=${tmpdir}/.git rev-parse --verify ${sha}`;
```
Comment (rclark) [inline scripts/cleanup.js]: By pulling out these long, complex statements and giving them a clear name, the flow of the rest of the application becomes a lot easier to read.

File: test/cleanup.test.js [hunk 5]
```diff
@@ -1,187 +1,284 @@
...
-const error = 'some error';
-const success = 'some success';
+const tmpdir = '/path/to/cloned/repo';
+
 const imagesNoToken = require(`${__dirname}/fixtures/imagesNoToken.test.json`);
 const imagesToken = require(`${__dirname}/fixtures/imagesToken.test.json`);
 const imageDetails = imagesNoToken.imageDetails.concat(imagesToken.imageDetails);
-const imageIds = [{ imageDigest: imagesNoToken.imageDetails[1].imageDigest }];
-
-test('mockSpawn', (t) => {
```
Comment (rclark) [inline test/cleanup.test.js]: I really don't like this pattern where the first test in a series sets up mocks used by latter tests. Its very important that each individual test can run in complete isolation from any other one.

File: test/cleanup.test.js [hunk 6]
```diff
@@ -1,187 +1,284 @@
...
-  let gitMergeCommitStdout;
-  let spawns = (command, args) => {
-    switch(command) {
-    case 'git':
-      if(args[3] === 'cccccccccccccccccccccccccccccccccccccccc' || args[3] === 'eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee') {
-        t.equals(args[0], `--git-dir=${process.argv[4]}/.git`);
```
Comment (rclark) [inline test/cleanup.test.js]: Furthermore -- what is `t` in this situation? It is in the closure for **this** test -- not the subsequent test that you run later which would fall into this block.

File: test/cleanup.test.js [hunk 7]
```diff
@@ -1,187 +1,284 @@
...
-  let stub = sinon.stub(process, 'exit');
-  let log = sinon.stub(console, 'log');
-  file.handleCb(error);
-  console.log.restore();
-  t.equal(stub.getCall(0).args[0], 1, 'error should exit 1');
-  t.equal(log.getCall(0).args[0], error, 'should log error');
-  process.exit.restore();
-  t.end();
-});
+test('getImages, error', (assert) => {
```
Comment (rclark) [inline test/cleanup.test.js]: OCD: I prefer `assert` to `t` -- just because its better at conveying what it is.

File: test/cleanup.test.js [hunk 8]
```diff
@@ -1,187 +1,284 @@
...
-  file.handleCb(null, success);
-  console.log.restore();
-  t.equal(stub.getCall(0).args[0], 0, 'success should exit 0');
-  t.equal(log.getCall(0).args[0], success, 'should log success');
-  process.exit.restore();
-  t.end();
-});
+  cleanup.getImages(region, repo, (err) => {
+    assert.ok(
+      stub.calledWith({ repositoryName: repo }),
```
Comment (rclark) [inline test/cleanup.test.js]: `.calledWith()` is a great part of sinon's API -- far easier to read than `stub.getCall(0).args[0]`.

**Comment by rclark** [comment]
I deployed a version of this where `MAX_IMAGES` was 1. I pushed a commit to a repo that had 10 generic and 1 tagged image in ECR. After ecs-conex ran, we were left with the tag image and the new image. All other images had been deleted, as anticipated.

File: scripts/cleanup.js (hunk 3 above)
Comment (arunasank) [reply scripts/cleanup.js]: woah. O.O

**Comment by arunasank** [comment]
Hey @rclark I love it when we get to work on something together, because your code reviews are always extensive, and turn my ugly code into something simple, and clear. : ) A long time ago, I used to really like this poem by Linda Pastan, called ["A New Poet"](https://wordsfortheyear.com/2015/05/18/a-new-poet-by-linda-pastan/), and every time, you review my code I am reminded of three lines in that poem - I've always thought it a little excessive to say, but I think I should. 🙂 :
```
And the words are so familiar,
so strangely new, words
you almost wrote yourself, if only
```

Except I'd replace "words" with "code". 😄 Thank you, for picking this up. All the changes make sense to me.

I saw that you tested this with an existing repository - I am also happy to pick up some more testing on `ecs-conex-test`, for more test cases, just to be more thorough, if this can wait a little longer. 

For example, I think making sure that all of these cases work, would be really useful.

* MAX_IMAGES = 10, MAX_PRIORITY_IMAGES = 5, say.

Generic | Priority | Sum | Example
----------|----------|--------|-----------|
< Max Images | < Priority Max | < Max Images | GENERIC = 5, PRIORITY = 3 => GENERIC = 5, PRIORITY = 3 |
< Max Images | < Priority Max | = Max Images | GENERIC = 6, PRIORITY = 4 => GENERIC - 6, PRIORITY = 4 |
< Max Images | < Priority Max | > Max Images | GENERIC = 7, PRIORITY = 4 => GENERIC - 6, PRIORITY = 4 |
< Max Images | > Priority Max | < Max Images | GENERIC = 3, PRIORITY = 6 => GENERIC - 3, PRIORITY = 6 |
< Max Images | > Priority Max | = Max Images | GENERIC = 3, PRIORITY = 7 => GENERIC - 3, PRIORITY = 7 |
< Max Images | > Priority Max | > Max Images | GENERIC = 3, PRIORITY = 8 => GENERIC - 2, PRIORITY = 8 |

And cases from all of the 27 possibilities. At the very least, I think we should at least test the cases, where the total number of images is less than the max, = max, > max(which you already tested.

Again, thank you for picking this up from the other branch, and opening a separate PR so that your changes are very visible. 🙇‍♀️

**Pull request merged**